### PR TITLE
Event storage

### PR DESCRIPTION
**Pull request by tanishqaggarwal**
This implements an "event log", i.e. multiple fields for storing the same kind of event. The idea behind this is that such a storage mechanism automatically stores a new event in the next available "slot" for storing an event, wrapping around stale data if necessary.

### CODE REVIEW

File: src/fsw/FCCode/Event.cpp [hunk 1]
```diff
@@ -19,6 +19,12 @@ Event::Event(const std::string& name,
     }
 }
 
+Event::Event(Event&& other) :
+    StateField<bool>(other.name(), true, false),
```
Comment (kkrol27) [inline src/fsw/FCCode/Event.cpp]: I'm assuming here that `other.name()` is a `std::string const &`? If so, everything looks good.

Comment (tanishqaggarwal) [reply src/fsw/FCCode/Event.cpp]: Yep

File: src/fsw/FCCode/UplinkCommon.cpp [hunk 2]
```diff
@@ -59,7 +59,7 @@ bool Uplink::_validate_packet(bitstream& bs)
     // if padding was not 0
     return (u8 != 0 || bits_consumed > 7) ? false : true;
 }
-#include <iostream>
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/UplinkCommon.cpp]: We should be extremely careful about leaving iostream as an include since it bloats up the code size by about 160 KB (15% of total memory.)

File: src/fsw/FCCode/EventStorage.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+void EventStorage::add_event_to_registry(StateFieldRegistry& registry) {
+    for(Event& e : sub_events) {
+        registry.add_readable_field(static_cast<ReadableStateFieldBase*>(&e));
+    }
+}
+
+size_t EventStorage::bitsize() const {
+    return sub_events[event_ptr].bitsize();
```
Comment (athena255) [inline src/fsw/FCCode/EventStorage.cpp]: Is sub_events[event_ptr] always guaranteed to have a bit_array?

File: test/test_fsw_event/test_fsw_event.cpp [hunk 4]
```diff
@@ -32,18 +33,17 @@ struct TestFixture {
...
 
     // Set values for test
     tf.control_cycle_count = 20;
     tf.data1_f.set(true);
     tf.data2_f.set(false);
-    
+
     // Verify that upon serialization, the values are written into the event's bitset in the way
     // that we would expect
     tf.event.signal();
```
Comment (athena255) [inline test/test_fsw_event/test_fsw_event.cpp]: Why do we copy ba to ccno instead of using the bit_array to ulong function?

File: src/fsw/FCCode/EventStorage.hpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#ifndef EVENT_STORAGE_HPP_
+#define EVENT_STORAGE_HPP_
+
```
Comment (athena255) [inline src/fsw/FCCode/EventStorage.hpp]: What is EventStorage suppose to do?

File: src/fsw/FCCode/Event.cpp [hunk 6]
```diff
@@ -19,6 +19,12 @@ Event::Event(const std::string& name,
...
+    data_fields(other.data_fields),
+    field_data(std::move(other.field_data)),
+    ccno(other.ccno) {}
+
```
Comment (athena255) [inline src/fsw/FCCode/Event.cpp]: in serialize(), is field_data guaranteed to have more capacity than 32 + field->bitsize()?

File: test/test_fsw_event/test_fsw_event.cpp [hunk 7]
```diff
@@ -65,10 +65,46 @@ void test_event() {
...
+    test_single_event(tf);
+    test_single_event(tf);
+    test_single_event(tf);
+}
+
 #ifdef DESKTOP
 int main() {
     UNITY_BEGIN();
     RUN_TEST(test_event);
+    RUN_TEST(test_event_storage);
```
Comment (athena255) [inline test/test_fsw_event/test_fsw_event.cpp]: I think it might be useful to test boundary cases. For example, test that upper limits of event storage fail elegantly when they should, and/or test with adding more than one event to registry.

File: src/fsw/FCCode/EventStorage.cpp (hunk 3 above)
Comment (tanishqaggarwal) [reply src/fsw/FCCode/EventStorage.cpp]: I think so, since they're all actual event objects.

File: src/fsw/FCCode/EventStorage.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                 const unsigned int& _ccno)
+{
+    assert(storage_size < 100); // So that the suffixed event count doesn't have more than 2 digits
+    sub_events.reserve(storage_size);
+    for(unsigned char i = 1; i <= storage_size; i++) {
+        char x[4];
+        memset(x, 0, 4);
+        x[0] = '.';
+        sprintf(x, ".%d", i);
+        sub_events.emplace_back(name + std::string(x),
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/fsw/FCCode/EventStorage.cpp]: Why should we do that?

File: src/fsw/FCCode/EventStorage.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+EventStorage::EventStorage(const std::string& name,
+                 const unsigned int storage_size,
+                 std::vector<ReadableStateFieldBase*>& _data_fields,
+                 const char* (*_print_fn)(const unsigned int, std::vector<ReadableStateFieldBase*>&),
+                 const unsigned int& _ccno)
+{
+    assert(storage_size < 100); // So that the suffixed event count doesn't have more than 2 digits
+    sub_events.reserve(storage_size);
+    for(unsigned char i = 1; i <= storage_size; i++) {
+        char x[4];
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/fsw/FCCode/EventStorage.cpp]: I think it's ok, it's a very small amount of memory. It's all stack-allocated anyway.

File: src/fsw/FCCode/EventStorage.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+EventStorage::EventStorage(const std::string& name,
+                 const unsigned int storage_size,
+                 std::vector<ReadableStateFieldBase*>& _data_fields,
+                 const char* (*_print_fn)(const unsigned int, std::vector<ReadableStateFieldBase*>&),
+                 const unsigned int& _ccno)
+{
+    assert(storage_size < 100); // So that the suffixed event count doesn't have more than 2 digits
+    sub_events.reserve(storage_size);
+    for(unsigned char i = 1; i <= storage_size; i++) {
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/fsw/FCCode/EventStorage.cpp]: Why? We only need `storage_size` spaces for the events.

File: src/fsw/FCCode/EventStorage.cpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        memset(x, 0, 4);
+        x[0] = '.';
+        sprintf(x, ".%d", i);
+        sub_events.emplace_back(name + std::string(x),
+            _data_fields, _print_fn, _ccno);
+    }
+}
+
+void EventStorage::add_event_to_registry(StateFieldRegistry& registry) {
+    for(Event& e : sub_events) {
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/fsw/FCCode/EventStorage.cpp]: We could do that.

[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/fsw/FCCode/EventStorage.cpp]: The pointers added to the registry are not const so we can't use const references.

File: test/test_fsw_event/test_fsw_event.cpp (hunk 4 above)
Comment (tanishqaggarwal) [reply test/test_fsw_event/test_fsw_event.cpp]: I could do that. Thanks!

File: src/fsw/FCCode/EventStorage.hpp (hunk 5 above)
Comment (tanishqaggarwal) [reply src/fsw/FCCode/EventStorage.hpp]: It's essentially a storage for an event log. It points to the next "free" event to store an event in.

File: src/fsw/FCCode/Event.cpp (hunk 6 above)
Comment (tanishqaggarwal) [reply src/fsw/FCCode/Event.cpp]: Yep because the constructor of Event() guarantees this

**Comment by tanishqaggarwal** [comment]
Closing while I make significant changes.
